src/charm.py

```diff
@@ -0,0 +1,256 @@
+#!/usr/bin/env python3
+
+import os
+import subprocess
+import socket
+import re
+import pwd
+import sys
+sys.path.append('lib')  # noqa
+
+from ops.charm import CharmBase, CharmEvents
+from ops.framework import EventBase, EventSource, StoredState
+from ops.main import main
+from ops.model import (
+    ActiveStatus,
+    BlockedStatus,
+    MaintenanceStatus,
+    WaitingStatus,
+    ModelError,
+)
+from cluster import CockroachDBCluster
+from interface_proxy_listen_tcp import ProxyListenTcpInterfaceProvides
+
+from jinja2 import Environment, FileSystemLoader
+from datetime import timedelta
+from time import sleep
+
+
+class CockroachStartedEvent(EventBase):
+    pass
+
+
+class ClusterInitializedEvent(EventBase):
+    def __init__(self, handle, cluster_id):
+        super().__init__(handle)
+        self.cluster_id = cluster_id
+
+    def snapshot(self):
+        return self.cluster_id
+
+    def restore(self, cluster_id):
+        self.cluster_id = cluster_id
+
+
+class CockroachDBCharmEvents(CharmEvents):
+    cockroachdb_started = EventSource(CockroachStartedEvent)
+    cluster_initialized = EventSource(ClusterInitializedEvent)
+
+
+class CockroachDBCharm(CharmBase):
+    on = CockroachDBCharmEvents()
+    state = StoredState()
+
+    COCKROACHDB_SERVICE = 'cockroachdb.service'
+    SYSTEMD_SERVICE_FILE = f'/etc/systemd/system/{COCKROACHDB_SERVICE}'
+    WORKING_DIRECTORY = '/var/lib/cockroach'
+    COCKROACH_INSTALL_DIR = '/usr/local/bin'
+    COCKROACH_BINARY_PATH = f'{COCKROACH_INSTALL_DIR}/cockroach'
+    COCKROACH_USERNAME = 'cockroach'
+    PSQL_PORT = 26257
+    HTTP_PORT = 8080
+
+    MAX_RETRIES = 10
+    RETRY_TIMEOUT = timedelta(milliseconds=125)
+
+    def __init__(self, framework, key):
+        super().__init__(framework, key)
+
+        self.state.set_default(is_started=False)
+
+        for event in (self.on.install,
+                      self.on.start,
+                      # self.on.upgrade_charm,
+                      self.on.config_changed,
+                      self.on.cluster_relation_changed,
+                      self.on.cockroachdb_started,
+                      self.on.proxy_listen_tcp_relation_joined):
+            self.framework.observe(event, self)
+
+        self.cluster = CockroachDBCluster(self, 'cluster')
+        self.tcp_load_balancer = ProxyListenTcpInterfaceProvides(self, 'proxy-listen-tcp')
+
+    def on_install(self, event):
+        try:
+            resource_path = self.model.resources.fetch('cockroach-linux-amd64')
+        except ModelError:
+            resource_path = None
+
+        if resource_path is None:
+            ARCHITECTURE = 'amd64'  # hard-coded until it becomes important
```

Comment:
Why all caps when this is a local variable?

What if resource_path is not None?  The variable will be unset and the logic below will break down, I think?

Reply:
> Why all caps when this is a local variable?

Just to indicate that it is meant to be a constant on purpose.
--

If a resource is not provided at the deployment time or later, and an attempt is made to retrieve it, a `ModelError` will be raised which will result in `None` being assigned to  `resource_path` (so we can just try and download a binary instead). Otherwise, it will be a valid file system path.